### PR TITLE
Use gather_nd in RNN-T gather_loss

### PR DESCRIPTION
**Pull request by antimora**
## Summary

- Replaces the chained `reshape` + `expand` + `gather` pattern in `RNNTLoss::gather_loss` with two `gather_nd` calls (#4709).
- Drops one fully-broadcast `[B, 1, U+1]` index materialization and one kernel launch per loss evaluation.
- Removes the now-orphaned `max_up1` parameter from the private signature and its single call site.

## Test plan

- [x] `cargo test -p burn-nn --lib loss::rnnt` - all 13 tests pass, including the 5 PyTorch numerical-parity cases (`basic_b1`, `batched_b2`, `variable_lengths_b3`, `mean_reduction`, `sum_reduction`).

Related cleanup ideas not in scope here are tracked in #4893.

### CODE REVIEW

File: crates/burn-nn/src/loss/rnnt.rs [hunk 1]
```diff
@@ -232,19 +233,17 @@ impl RNNTLoss {
...
-
-        // Gather blank prob at (T_b, U_b)
-        let t_exp = t_idx.reshape([b, 1, 1]).expand([b, 1, max_up1]);
-        let lpb_t = lpb.clone().gather(1, t_exp).squeeze_dim::<2>(1);
-        let lpb_tu = lpb_t.gather(1, u_idx.reshape([b, 1])).squeeze_dim::<1>(1);
+        let alpha_tu: Tensor<B, 1> =
+            alpha.gather_nd(Tensor::stack::<2>(vec![b_idx.clone(), u_idx.clone()], 1));
+        let lpb_tu: Tensor<B, 1> = lpb
+            .clone()
+            .gather_nd(Tensor::stack::<2>(vec![b_idx, t_idx, u_idx], 1));
```
[the remark this answers is not in the document]
Comment (antimora) [reply crates/burn-nn/src/loss/rnnt.rs:250]: Good catch — fixed in 101a8d3. Anchored the coord dtype on `u_idx.dtype()`, created `b_idx` with `(&device, int_dtype)`, and added a `t_idx.cast(int_dtype)` so all three coordinate tensors share a dtype before `Tensor::stack`/`cat`. The `Cast<B, Int>` impl short-circuits when dtypes already match (`cast.rs:62`), so this is a no-op when the caller passes consistent lengths and only inserts a real conversion if `logit_lengths` and `target_lengths` actually differ. Aligns with the int-dtype policy from #4780.

**Review by laggui — APPROVED** [review]
